lib/lz4.c

```diff
@@ -1178,8 +1178,25 @@ LZ4_FORCE_INLINE int LZ4_decompress_generic(
         const BYTE* match;
         size_t offset;
 
-        /* get literal length */
         unsigned const token = *ip++;
+
+        /* shortcut for common case */
+        /* this shortcut was tested on x86 and x64, where it improves decoding speed.
+         * it has not yet been benchmarked on ARM, Power, mips, etc. */
+        if (((ip + 14 + 2 <= iend) & (op + 14 + 16 <= oend))
+          & ((token < 15*16) & ((token & 0xF) <= 12)) ) {
+            size_t const ll = token >> ML_BITS;
+            size_t const off = LZ4_readLE16(ip+ll);
+            const BYTE* const matchPtr = op + ll - off;  /* pointer underflow risk ? */
+            if ((off >= 16) & (matchPtr >= lowPrefix)) {
+                size_t const ml = (token & 0xF) + MINMATCH;
+                memcpy(op, ip, 16); op += ll; ip += ll + 2 /*offset*/;
```

Comment:
How does the speed compare to wildcopy, or just two `memcpy(op, ip, 8);`?

Reply:
not tested yet.
`memcpy(,,16)` gives a chance to use an SSE register.
I expect it to nicely degrade to two `memcpy(,,8)` when such register is not available.

On systems without 16-bytes registers, `memcpy(,,8)` for literals could prove a better choice, though it highly depends on the distribution of literal lengths values.
If we assume that most literal lengths are `<= 2`, an 8-bytes copy is likely enough, and costs less work for system limited to 64-bits memory transfers.
But if the distribution is a bit wider, literal lengths from `9` to `14` become no longer eligible to shortcut, making the shortcut test less predictable. The impact on the branch predictor might be more problematic than the extra 8-bytes copy.

Reply:
I just tested a variant limiting literal lengths to 8, in order to use a `memcpy(,,8)` for literals.
Its speed is _significantly_ worse, even worse than without the patch.

It seems that making the branch predictor reliable is paramount to performance.

Reply:
Ahh sorry, I meant `memcpy(op, ip, 8); memcpy(op + 8, ip + 8, 8);` instead of `memcpy(op, ip, 16);`. Looking at the godbolt output, the difference will be using SSE vs not using SSE. I suspect SSE will win, especially since that is what the compiler outputs.

If 16 does win by a good amount, could wildcopy be using 16 bytes? Or would the increased overlap range where we have to special case the copy be a large slowdown.

Reply:
moving all `wildcopy()` operations `16` is more complex to implement.
there are a number of tests which depend on `wildcopy()` over-writing to be limited to 8,
and the overlapping match code also depends on it.

I made a test on this scenario a long time ago, the result was so-so, not bad but not terribly good, and seemed not worth added code complexity.
It could also be that the modifications were not done well enough, masking some of the potential benefits.